clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1657,6 +1658,87 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
   using SyclKernelFieldHandler::leaveStruct;
 };
 
+class SyclKernelNumArgsChecker : public SyclKernelFieldHandler {
+  SourceLocation KernelLoc;
+  unsigned NumOfParams = 0;
+
+  void addParam() { NumOfParams++; }
+
+  bool handleSpecialType(QualType FieldTy) {
+    const auto *RecordDecl = FieldTy->getAsCXXRecordDecl();
+    assert(RecordDecl && "The accessor/sampler must be a RecordDecl");
+    CXXMethodDecl *InitMethod = getMethodByName(RecordDecl, InitMethodName);
+    assert(InitMethod && "The accessor/sampler must have the __init method");
+    unsigned NumOfParams = InitMethod->getNumParams();
+    for (unsigned I = 0; I < NumOfParams; ++I)
```

Comment:
I wonder slightly if this is 'too cute'.  Since you're essentially just doing: this.NumOfParams+= NumOfParams;

Also, the name shadowing is unfortuante.

Reply:
Oh, I didn't mean name shadowing.
Will removing of `addParam` function and doing change of `this.NumOfParams` every time make it not 'too cute'?

Reply:
That would definitely avoid being at all 'cute'.  I get why you did it (to match the function in the other handler more closely), but I'm on the fence as to its value over just doing NumofParams += InitMethod->getNumParams();.

Alternatively, you could consider having the addParam function take a parameter for how many to add (perhaps defaulted to 1) if you think it increases readability.

Reply:
Why do we need this loop?  Wouldn't:
NumOfParams = InitMethod->getNumParams();
be equivalent?

Reply:
Rather '+=' instead of '='.

Reply:
Okay, done